IDEA/src/main/java/ksp/kos/ideaplugin/reference/context/LocalContext.kt

```diff
@@ -1,169 +1,97 @@
-package ksp.kos.ideaplugin.reference.context;
+package ksp.kos.ideaplugin.reference.context
 
-import ksp.kos.ideaplugin.psi.KerboScriptNamedElement;
-import ksp.kos.ideaplugin.reference.ReferableType;
-import ksp.kos.ideaplugin.reference.Reference;
-import org.jetbrains.annotations.NotNull;
-import org.jetbrains.annotations.Nullable;
-
-import java.util.*;
+import ksp.kos.ideaplugin.psi.KerboScriptNamedElement
+import ksp.kos.ideaplugin.reference.ReferableType
+import ksp.kos.ideaplugin.reference.Reference
+import java.util.*
+import kotlin.collections.HashMap
 
 /**
  * Created on 04/10/16.
  *
  * @author ptasha
  */
-public class LocalContext {
-    protected final LocalContext parent;
-    private final Map<ReferableType, Map<String, Duality>> declarations = new HashMap<>();
-
-    private final List<ReferenceResolver<LocalContext>> resolvers;
-
-    public LocalContext(LocalContext parent) {
-        this(parent, createResolvers());
-    }
-
-    protected LocalContext(LocalContext parent, List<ReferenceResolver<LocalContext>> resolvers) {
-        this.parent = parent;
-        this.resolvers = resolvers;
-    }
-
-    public static List<ReferenceResolver<LocalContext>> createResolvers() {
-        List<ReferenceResolver<LocalContext>> resolvers = new ArrayList<>();
-        resolvers.add(new LocalResolver());
-        resolvers.add(new ParentResolver());
-        return resolvers;
-    }
+open class LocalContext protected constructor(
+    val parent: LocalContext?,
+    private val resolvers: List<ReferenceResolver<LocalContext>>,
+) {
+    private val declarations: MutableMap<ReferableType, MutableMap<String, Duality>> = HashMap()
 
-    public LocalContext getParent() {
-        return parent;
-    }
+    constructor(parent: LocalContext?) : this(parent, createResolvers())
```

Comment:
I think this can be replaced with a default parameter in the first constructor, like 

```kotlin
private val resolvers: List<ReferenceResolver<LocalContext>> = createResolvers()
```

Which, in my opinion, is more readable.

Reply:
Oh nice, I like that better too! It does mean it can't be `protected `anymore due to this usage:
https://github.com/KSP-KOS/EditorTools/blob/377becfc75004cd1d8322fe00e2e49f22ba5f36b/IDEA/src/main/java/ksp/kos/ideaplugin/reference/PsiFileContext.java#L29
but on the whole I think that's a worthwhile tradeoff🙂